File: src/Renderer.cpp
#include "../include/Renderer.h"
#include <cmath>

#define PI acos(-1)

void Renderer::render(Solver &solver)
{
    // render bounding box
    RectBounds bounds = solver.getBounds();

    glColor3f(1.0, 1.0, 1.0);

    glBegin(GL_QUADS);
    glVertex2d(bounds.left, bounds.down);
    glVertex2d(bounds.right, bounds.down);
    glVertex2d(bounds.right, bounds.up);
    glVertex2d(bounds.left, bounds.up);
    glEnd();

    //=====================================================================

    std::vector<Circle>& objects = solver.getObjects();
    float vectorScale = 0.02;
    int lineWidth = 3;

    for (int i = 0; i < objects.size(); i++)
    {
        Circle& obj = objects[i];
        float red = obj.colour[0];
        float green = obj.colour[1];
        float blue = obj.colour[2];

        // render objects
        glColor3f(red, green, blue);

        glBegin(GL_POLYGON);
        for (int vertex = 0; vertex < RESOLUTION; vertex++)
        {
            float vertexX = obj.radius * cos(vertex * 2 * PI / RESOLUTION) + obj.pos.x();
            float vertexY = obj.radius * sin(vertex * 2 * PI / RESOLUTION) + obj.pos.y();
            glVertex2d(vertexX, vertexY);
        }
        glEnd();

        // render object velocity vectors
        glColor3f(1 - red, 1 - green, 1 - blue);
        glLineWidth(lineWidth);

        glBegin(GL_LINES);
        glVertex2d(obj.pos.x(), obj.pos.y());
        glVertex2d(obj.pos.x() + vectorScale * obj.vel.x(), obj.pos.y() + vectorScale * obj.vel.y());
        glEnd();
    }
}

File: py/Renderer.py
from math import *
from OpenGL.GL import *
from Solver import *

class Renderer:

    RESOLUTION = 20 # no. of sides on circle approximation

    @classmethod
    def render(cls, solver: Solver):

        # render bounding box
        glColor(1.0, 1.0, 1.0)  # white

        glBegin(GL_QUADS)
        glVertex2d(solver.BOUNDS.left, solver.BOUNDS.down)
        glVertex2d(solver.BOUNDS.right, solver.BOUNDS.down)
        glVertex2d(solver.BOUNDS.right, solver.BOUNDS.up)
        glVertex2d(solver.BOUNDS.left, solver.BOUNDS.up)
        glEnd()

        # render objects
        for object in solver.objects:

            glColor(object.colour[0], object.colour[1], object.colour[2])

            glBegin(GL_POLYGON)
            for vertex in range(cls.RESOLUTION):
                
                # draw circle around object position
                vertex_x = object.radius * cos(vertex * 2 * pi / cls.RESOLUTION) + object.pos.x
                vertex_y = object.radius * sin(vertex * 2 * pi / cls.RESOLUTION) + object.pos.y

                glVertex2d(vertex_x, vertex_y)
            glEnd()

        # render object velocity vectors
        vector_scale = 0.02
        line_width = 3

        for object in solver.objects:

            glColor(1-object.colour[0], 1-object.colour[1], 1-object.colour[2])
            glLineWidth(line_width)

            glBegin(GL_LINES)
            glVertex2d(object.pos.x, object.pos.y)
            glVertex2d(object.pos.x + vector_scale * object.vel.x, object.pos.y + vector_scale * object.vel.y)
            glEnd()

        # render grid
        line_width = 1

        glColor(0, 0, 0)
        glLineWidth(line_width)

        # horizonal lines
        for i in range(solver.grid.cell_size, solver.WINDOW_H, solver.grid.cell_size):
            glBegin(GL_LINES)
            glVertex2d(0, i)
            glVertex2d(solver.WINDOW_W, i)
            glEnd()
        # vertical lines
        for i in range(solver.grid.cell_size, solver.WINDOW_W, solver.grid.cell_size):
            glBegin(GL_LINES)
            glVertex2d(i, 0)
            glVertex2d(i, solver.WINDOW_H)
            glEnd()